src/Features/CSharp/Portable/EditAndContinue/CSharpEditAndContinueAnalyzer.cs

```diff
@@ -1041,6 +1114,59 @@ internal override bool HasBackingField(SyntaxNode propertyOrIndexerDeclaration)
         internal override bool IsDeclarationWithInitializer(SyntaxNode declaration)
             => declaration is VariableDeclaratorSyntax { Initializer: not null } || declaration is PropertyDeclarationSyntax { Initializer: not null };
 
+        internal override bool IsRecordPrimaryConstructorParameter(SyntaxNode declaration)
+            => declaration is ParameterSyntax { Parent: ParameterListSyntax { Parent: RecordDeclarationSyntax } };
+
+        internal override bool IsPropertyDeclarationMatchingPrimaryConstructorParameter(SyntaxNode declaration)
+            => declaration is PropertyDeclarationSyntax { Parent: RecordDeclarationSyntax record, Identifier: { ValueText: var name } }
+            && record.ParameterList is not null
+            && record.ParameterList.Parameters.Any(p => p.Identifier.ValueText.Equals(name));
+
+        internal override bool IsPropertyAccessorDeclarationMatchingPrimaryConstructorParameter(SyntaxNode declaration, out bool isFirstAccessor)
+        {
+            isFirstAccessor = false;
+            if (declaration is AccessorDeclarationSyntax { Parent: AccessorListSyntax { Parent: PropertyDeclarationSyntax property } list } &&
+                IsPropertyDeclarationMatchingPrimaryConstructorParameter(property))
+            {
+                isFirstAccessor = list.Accessors[0] == declaration;
+                return true;
+            }
+            return false;
+        }
+
+        internal override IEnumerable<ISymbol> GetRecordUpdatedSynthesizedMembers(Compilation compilation, INamedTypeSymbol record)
+        {
+            // All methods that are updated have well known names, and calling GetMembers(string) is
+            // faster than enumerating.
+
+            // When a new field or property is added the PrintMembers, Equals(R) and GetHashCode() methods are updated
+            // We don't need to worry about Deconstruct because it only changes when a new positional parameter
+            // is added, and those are rude edits (due to adding a constructor parameter).
+            // We don't need to worry about the constructors as they are reported elsewhere.
+            // We have to use SingleOrDefault and check IsImplicitlyDeclared because the user can provide their
+            // own implementation of these methods, and edits to them are handled by normal processing.
+            yield return record.GetMembers(WellKnownMemberNames.PrintMembersMethodName)
+                .OfType<IMethodSymbol>()
+                .SingleOrDefault(m =>
```

Comment:
>SingleOrDefault [](start = 17, length = 15)

This will throw if there are two matching members, which may happen in error cases - it'd be good to add a test that has these members duplicated.

Reply:
Can we avoid returning null here - I wonder why doesn't the compiler flag this since the return type does not allow nullable ISymbol

---
In reply to: [609973366](https://github.com/dotnet/roslyn/pull/51926#discussion_r609973366) [](ancestors = 609973366)

Reply:
> I wonder why doesn't the compiler flag this since the return type does not allow nullable ISymbol

The API was annotated starting in net5.0. I think the TFM used is lower than that, hence, the API is not annotated in it.

Reply:
> This will throw if there are two matching members, which may happen in error cases

Unless I'm misunderstanding, the `m.IsImplicitlyDeclared` part of the predicate means it would only throw if there are two _compiler generated_ members. Is that something we need to worry about? Seems a hell of a compiler bug, and not something I can create a test for.

When a user explicitly implements one of these members we don't need to automatically issue edits, because if they change the contents of their methods then the normal edit creation would have dealt with those methods. For example the `Record_AddProperty_NotPrimary_WithExplicitMembers` test illustrates this by not issue edits for `PrintMembers` etc.

I've rejiggered this to not return nulls though.

Reply:
You are right. I missed the IsImplicitlyDeclared condition. Looks good.